ogr/services/gitlab/service.py

```diff
@@ -128,3 +128,53 @@ def project_create(
         return GitlabProject(
             repo=repo, namespace=namespace, service=self, gitlab_repo=new_project
         )
+
+    def list_projects(
+        self,
+        namespace: str = None,
+        user: str = None,
+        search_pattern: str = None,
+        language: str = None,
+    ) -> List[GitProject]:
+
+        if namespace:
+            group = self.gitlab_instance.groups.get(namespace)
+            projects = group.projects.list(all=True)
+        if user:
+            u = self.gitlab_instance.users.list(username=user)[0]
+            projects = u.projects.list(all=True)
+
+        gitlab_projects: List[GitProject]
+
+        if language:
+            # group.projects.list gives us a GroupProject instance
+            # in order to be able to filter by language we need Project instance
+            projects_filterable_by_language = [
+                self.gitlab_instance.projects.get(item.attributes["id"])
+                for item in projects
+            ]
+
+            gitlab_projects = [
+                GitlabProject(
+                    repo=project.attributes["path"],
+                    namespace=project.attributes["namespace"]["full_path"],
+                    gitlab_repo=project,
+                    service=self,
+                )
+                for project in projects_filterable_by_language
+                if language
+                in project.languages().keys()  # projects.languages() returns a dict
+            ]
+
+        else:
+            gitlab_projects = [
+                GitlabProject(
+                    repo=project.attributes["path"],
+                    namespace=project.attributes["namespace"]["full_path"],
+                    gitlab_repo=project,
+                    service=self,
+                )
+                for project in projects
+            ]
```

Comment:
This if statement is pretty huge and the core part is duplicated.

What about something like this?


```suggestion
        if language:
            # group.projects.list gives us a GroupProject instance
            # in order to be able to filter by language we need Project instance
            projects_to_convert = [
                self.gitlab_instance.projects.get(item.attributes["id"])
                for item in projects
                if language
                in project.languages().keys()
            ]
        else:
            projects_to_convert = projects

        gitlab_projects = [
            GitlabProject(
                repo=project.attributes["path"],
                namespace=project.attributes["namespace"]["full_path"],
                gitlab_repo=project,
                service=self,
            )
            for project in projects_to_convert
        ]
```

Reply:
Yeah this totally makes sense. I will update the code accordingly.